staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package feature
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+)
+
+var (
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_31 = version.MustParseSemantic("3.4.31")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_13 = version.MustParseSemantic("3.5.13")
+
+	// DefaultFeatureSupportChecker is a shared global etcd FeatureSupportChecker.
+	DefaultFeatureSupportChecker FeatureSupportChecker = newDefaultFeatureSupportChecker()
+)
+
+// FeatureSupportChecker to define Supports functions.
+type FeatureSupportChecker interface {
+	Supports(feature storage.Feature) (bool, error)
+	// This method works with etcd client to recalculate feature support and caches it.
+	ClientSupports(ctx context.Context, c client, feature storage.Feature) (bool, error)
```

Comment:
> I don't fully understand about why we should make ClientSupports private? read comments below also i don't know how to make a global variable of DefaultFeatureSupportChecker with client as a parameter in constructor, i want to make this better but need some more information thanks.
> In previous PR we needed to call ClientSupports on etcd3/store.go file and probably save result for each client in feature module.


So here is hand-wavy idea that I would like to be explored:
1) We don't expose `DefaultFeatureSupportChecker", but instead expose
    `func GetDefaultFeatulreSupportChecker() *FeatureSupportChecker {`
2) the callers of that function need to be prepared that this is not yet initialized
3) the initialization is done from etcd3 package that will initialize it

The above means that we probably will need to move it to etcd3 package or alternatiely expose a public InitializeDefaultChecker function, but I think both of these options are actually better.

Reply:
@serathius for your thoughts

Reply:
We chatted offline with Marek about that and we set up on sth else meaning:
- ClientSupports isn't a getter, it's rather a setter
- we should rename it and change its signature to make it clear it's a setter, to sth like

`SetForClient(context, client, feature) error`  [better name would be great]

And we're both ok with this apprach as it makes it clear what is what.

Reply:
Sorry for late response, thanks you, it's more clear this way, let me clean the parts.